[FSDP] Fix wrapped module changing after ctor

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#87837 [FSDP] Fix wrapped module changing after ctor**
* #87812 [FSDP] ufmt FSDP test
* #87811 [FSDP] ufmt /fsdp

Recently, I retired `FlattenParamsWrapper`, which meant that FSDP registers its `FlatParameter` on the wrapped module instead of the `FlattenParamsWrapper` instance. This is only relevant for `use_orig_params=False`.

If the user changes an FSDP instance's wrapped module after the FSDP constructor, then the `FlatParameter` is no longer registered on the wrapped module. This can cause issues for full state dict, which checks if the `FlatParameter` is currently registered as an early return condition for `rank0_only=True`.

The solution in this PR is to re-establish the wrapped module in `_lazy_init()`, de-registering from the old wrapped module and re-registering to the new wrapped module, where the assumption is that the user should not modify the module structure upon `_lazy_init()`.

The direct access to the private attribute `_parameters` from `nn.Module` is not ideal, but we already rely on it for the dynamic `FlatParameter` registration. The tradeoff is whether we want an additional `nn.Module` wrapper (`FlattenParamsWrapper`) and use `delattr` plus a singleton list to do the dynamic registration or we want to access `_parameters`. If this becomes a problem, we can work with Core team on a solution.

Differential Revision: [D40799962](https://our.internmc.facebook.com/intern/diff/D40799962)

Commit:
[FSDP] Fix wrapped module changing after ctor

[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix wrapped module changing after ctor"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix wrapped module changing after ctor"


Recently, I retired `FlattenParamsWrapper`, which meant that FSDP registers its `FlatParameter` on the wrapped module instead of the `FlattenParamsWrapper` instance. This is only relevant for `use_orig_params=False`.

If the user changes an FSDP instance's wrapped module after the FSDP constructor, then the `FlatParameter` is no longer registered on the wrapped module. This can cause issues for full state dict, which checks if the `FlatParameter` is currently registered as an early return condition for `rank0_only=True`.

The solution in this PR is to re-establish the wrapped module in `_lazy_init()`, de-registering from the old wrapped module and re-registering to the new wrapped module, where the assumption is that the user should not modify the module structure upon `_lazy_init()`.

The direct access to the private attribute `_parameters` from `nn.Module` is not ideal, but we already rely on it for the dynamic `FlatParameter` registration. The tradeoff is whether we want an additional `nn.Module` wrapper (`FlattenParamsWrapper`) and use `delattr` plus a singleton list to do the dynamic registration or we want to access `_parameters`. If this becomes a problem, we can work with Core team on a solution.

[ghstack-poisoned]

fegin commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -489,6 +499,91 @@ def __init__(self, rank):
+            offload_to_cpu=False,
+            checkpoint_impl=CheckpointImpl.NO_REENTRANT,
+        )
+        check_fn = lambda submodule: isinstance(submodule, nn.Linear)
```
lambda should be anonymous

awgu replied:
Could you explain what the difference between anonymous and lambda is?

fegin commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -220,9 +234,10 @@ def _validate_state_dict_contents(
     @skip_if_lt_x_gpu(2)
     @parametrize("state_dict_type", _UNFLATTENED_STATE_DICT_IMPLS)
-    @parametrize("checkpoint_wrap", ["first", "second", "both"])
+    @parametrize("checkpoint_wrap", ["source", "dest", "both", "source_after_wrap"])
```
Curious, if we have `both_after_wrap`, then with rank0_only_and_offload being False, would it be possible to reproduce the loading stuck error?

awgu replied:
`"both_after_wrap"` with `rank0_only_and_offload=False` produces a different error:
```
RuntimeError: Error(s) in loading state_dict for FullyShardedDataParallel:
	While copying the parameter named "_fsdp_wrapped_module.0._fsdp_wrapped_module._checkpoint_wrapped_module._flat_param", whose dimensions in the model are torch.Size([100]) and whose dimensions in the checkpoint are torch.Size([100]), an exception occurred : ('CUDA error: invalid argument\nCUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
```
However, since it is still an error, I added the `"both_after_wrap"` option to the unit test. We still need to figure out how to reproduce the load error.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1884,6 +1884,41 @@ def _lazy_init(self) -> None:
+                        if flat_param is param:
+                            target_submodule = submodule
+                            target_name = param_name
+                            break
```
This only break the inner loop. Although there is only target_submodule, we should also break the outer loop to make the code less confusion.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1884,6 +1884,41 @@ def _lazy_init(self) -> None:
+                        f"these are invalid behavior. {fsdp_module}"
+                    )
+                if target_submodule is not fsdp_module.module:
+                    fsdp_module._register_flat_param()
```
Can this line be moved after line 1912, the if block in line 1901?

awgu replied:
Yes, good suggestion! I originally did not raise `RuntimeError` in the case that `target_submodule is None` and tried to still re-register, which was why this re-registration was here below. However, I realized that something would likely break soon after, so I chose to raise `RuntimeError`.

Commit:
Update on "[FSDP] Fix wrapped module changing after ctor"


Recently, I retired `FlattenParamsWrapper`, which meant that FSDP registers its `FlatParameter` on the wrapped module instead of the `FlattenParamsWrapper` instance. This is only relevant for `use_orig_params=False`.

If the user changes an FSDP instance's wrapped module after the FSDP constructor, then the `FlatParameter` is no longer registered on the wrapped module. This can cause issues for full state dict, which checks if the `FlatParameter` is currently registered as an early return condition for `rank0_only=True`.

The solution in this PR is to re-establish the wrapped module in `_lazy_init()`, de-registering from the old wrapped module and re-registering to the new wrapped module, where the assumption is that the user should not modify the module structure upon `_lazy_init()`.

The direct access to the private attribute `_parameters` from `nn.Module` is not ideal, but we already rely on it for the dynamic `FlatParameter` registration. The tradeoff is whether we want an additional `nn.Module` wrapper (`FlattenParamsWrapper`) and use `delattr` plus a singleton list to do the dynamic registration or we want to access `_parameters`. If this becomes a problem, we can work with Core team on a solution.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix wrapped module changing after ctor"


Recently, I retired `FlattenParamsWrapper`, which meant that FSDP registers its `FlatParameter` on the wrapped module instead of the `FlattenParamsWrapper` instance. This is only relevant for `use_orig_params=False`.

If the user changes an FSDP instance's wrapped module after the FSDP constructor, then the `FlatParameter` is no longer registered on the wrapped module. This can cause issues for full state dict, which checks if the `FlatParameter` is currently registered as an early return condition for `rank0_only=True`.

The solution in this PR is to re-establish the wrapped module in `_lazy_init()`, de-registering from the old wrapped module and re-registering to the new wrapped module, where the assumption is that the user should not modify the module structure upon `_lazy_init()`.

The direct access to the private attribute `_parameters` from `nn.Module` is not ideal, but we already rely on it for the dynamic `FlatParameter` registration. The tradeoff is whether we want an additional `nn.Module` wrapper (`FlattenParamsWrapper`) and use `delattr` plus a singleton list to do the dynamic registration or we want to access `_parameters`. If this becomes a problem, we can work with Core team on a solution.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix wrapped module changing after ctor"


Recently, I retired `FlattenParamsWrapper`, which meant that FSDP registers its `FlatParameter` on the wrapped module instead of the `FlattenParamsWrapper` instance. This is only relevant for `use_orig_params=False`.

If the user changes an FSDP instance's wrapped module after the FSDP constructor, then the `FlatParameter` is no longer registered on the wrapped module. This can cause issues for full state dict, which checks if the `FlatParameter` is currently registered as an early return condition for `rank0_only=True`.

The solution in this PR is to re-establish the wrapped module in `_lazy_init()`, de-registering from the old wrapped module and re-registering to the new wrapped module, where the assumption is that the user should not modify the module structure upon `_lazy_init()`.

The direct access to the private attribute `_parameters` from `nn.Module` is not ideal, but we already rely on it for the dynamic `FlatParameter` registration. The tradeoff is whether we want an additional `nn.Module` wrapper (`FlattenParamsWrapper`) and use `delattr` plus a singleton list to do the dynamic registration or we want to access `_parameters`. If this becomes a problem, we can work with Core team on a solution.

[ghstack-poisoned]

zhaojuanmao commented:
once we decided to move checkpoint wrapper to hook based, we may think about whether we want to support this use case, and maybe just error out and say changing module structures after FSDP wrapping is not supported

awgu commented:
@pytorchbot merge

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40799962).
